### DISCUSSION TITLE
introduce accessors to the state of the buf_ring

### DISCUSSION THREAD

**Discussion by romange** [axboe/liburing#1020] [Ideas]
similarly to `io_uring_cq_ready` could be nice to have a method returning number of buffers unclaimed yet by the kernel, i.e. those that need to be released by `io_uring_buf_ring_advance`. One possible application for that - for tests, to make sure the application code has released back the buffers it consumed.

**Comment by axboe** [comment]
Makes sense indeed.

**Reply by axboe** [reply]
We'd need kernel side support for this as well, as the kernel side consumption head isn't visible to userspace.

**Reply by axboe** [reply]
See:

https://lore.kernel.org/io-uring/7fb42968-17fd-47d8-b2cd-55a3382f1540@kernel.dk/

for the kernel side, and then see:

https://git.kernel.dk/cgit/liburing/commit/?h=pbuf-status&id=0769f32810d546395111eb1c601739259ffe0b00

for some basic liburing support, along with a test case.

**Comment by romange** [comment]
Thanks! What version it's gonna be merged to?

**Reply by axboe** [reply]
Could be as early as 6.8 on the kernel side, and then the liburing side would be in the next release of that. Pending review and discussion.

**Reply by romange** [reply]
Thank you.

**Reply by axboe** [reply]
This has landed in the 6.8 merge window, and will be part of the 6.8 kernel release. I just now merged the liburing support as well, which will land in the 2.6 release.

**Comment by romange** [comment]
Hi Jens, thank you!
